mountpoint-s3/src/data_cache/mod.rs

```diff
@@ -0,0 +1,54 @@
+//! Traits and types for data caching.
+//!
+//! The data cache aims to reduce repeated fetches of S3 object content,
+//! reducing both the number of requests as well as the latency for the reads.
+//! Ultimately, this means reduced cost in terms of S3 billing as well as compute time.
+
+mod in_memory_data_cache;
+
+use std::ops::Range;
+
+use thiserror::Error;
+
+pub use crate::prefetch::checksummed_bytes::ChecksummedBytes;
+
+/// Indexes blocks within a given object.
+pub type BlockIndex = u64;
+
+/// Errors returned by operations on a [DataCache]
+#[derive(Debug, Error)]
+pub enum DataCacheError {
+    /// It was not possible to read from the cache
+    #[error("Failed reading or writing from cache: {0}")]
+    IoFailure(#[source] anyhow::Error),
+}
+
+pub type DataCacheResult<Value> = Result<Value, DataCacheError>;
+
+/// Cache data with a checksum identified by some [Key].
+///
+/// The underlying cache is divided into blocks of equal size.
+///
+/// TODO: Deletion and eviction of cache entries.
+/// TODO: Some version information (ETag) independent from [Key] to allow smarter eviction?
+pub trait DataCache<Key> {
+    /// Get block of data from the cache for the given [Key] and [BlockIndex], if available.
+    ///
+    /// Operation may fail due to errors, or return [None] if the block was not available in the cache.
+    fn get_block(&self, cache_key: &Key, block_idx: BlockIndex) -> DataCacheResult<Option<&ChecksummedBytes>>;
```

Comment:
Two thoughts here:
- Does the client decide how block indexing works (and the implementation treats them as opaque), or do the implementation and client need to agree on how that works somehow?
- If it's the former, can they just be part of the generic `Key` type?

Reply:
>Does the client decide how block indexing works (and the implementation treats them as opaque), or do the implementation and client need to agree on how that works somehow?

I hadn't fully thought this through yet. I think for now we will move towards the implementation knowing about "blocks" and surfacing things like the cache's block size in the trait. Client will need to use that information to calculate things such as:

- If I want to fetch some data, what ranges should I use for full blocks?
- If I read from blocks, what do I need to trim at the start and end?

I'm hoping we can move that logic into the DataCache trait later, but for now I'm not sure exactly how it'll look like.